File: src/micapolos/tata8/Position.java
package micapolos.tata8;

public final class Position {
  public float x;
  public float y;

  public Position() {}

  public Position(float x, float y) {
    this.x = x;
    this.y = y;
  }

  public void setZero() {
    set(0, 0);
  }

  public void set(float x, float y) {
    this.x = x;
    this.y = y;
  }

  public void set(Position vector) {
    set(vector.x, vector.y);
  }

  public void add(float x, float y) {
    set(this.x + x, this.y + y);
  }

  public void add(Speed speed) {
    set(x + speed.x, y + speed.y);
  }

  public void setElastic(Position position) {
    set(Math.elastic(x, position.x), Math.elastic(y, position.y));
  }

  @Override
  public String toString() {
    return "vector(" + x + ", " + y + ")";
  }
}

File: src/micapolos/zexy2/runtime/Position.kt
package micapolos.zexy2.runtime

class Position<T: Number>(val x: Value<T>, val y: Value<T>)

fun position(x: Double, y: Double) = Position(x.value, y.value)
fun position(x: Double, y: Value<Double>) = Position(x.value, y)
fun position(x: Value<Double>, y: Double) = Position(x, y.value)
fun position(x: Value<Double>, y: Value<Double>) = Position(x, y)
